sw/airborne/firmwares/rotorcraft/stabilization/stabilization_attitude_quat_indi.c

```diff
@@ -0,0 +1,322 @@
+  u_act_dyn.p = u_act_dyn.p + STABILIZATION_INDI_ACT_DYN_P*( u_in.p - u_act_dyn.p);
+  u_act_dyn.q = u_act_dyn.q + STABILIZATION_INDI_ACT_DYN_Q*( u_in.q - u_act_dyn.q);
+  u_act_dyn.r = u_act_dyn.r + STABILIZATION_INDI_ACT_DYN_R*( u_in.r - u_act_dyn.r);
+
+  //sensor filter
+  stabilization_indi_second_order_filter(&u_act_dyn, &udotdot, &udot, &indi_u, STABILIZATION_INDI_FILT_OMEGA, STABILIZATION_INDI_FILT_ZETA, STABILIZATION_INDI_FILT_OMEGA_R);
+
+  //Don't increment if thrust is off
+  if(stabilization_cmd[COMMAND_THRUST]<300) {
+    FLOAT_RATES_ZERO(indi_u);
+    FLOAT_RATES_ZERO(indi_du);
+    FLOAT_RATES_ZERO(u_act_dyn);
+    FLOAT_RATES_ZERO(u_in);
+    FLOAT_RATES_ZERO(udot);
+    FLOAT_RATES_ZERO(udotdot);
+  }
+  else {
+#if STABILIZATION_INDI_USE_ADAPTIVE
+    lms_estimation();
+#endif
+  }
+
+  /*  INDI feedback */
+  indi_commands[COMMAND_ROLL] = u_in.p;
+  indi_commands[COMMAND_PITCH] = u_in.q;
+  indi_commands[COMMAND_YAW] = u_in.r;
+}
+
+void stabilization_attitude_run(bool_t enable_integrator) {
+
+  /* Propagate the second order filter on the gyroscopes */
+  struct FloatRates *body_rates = stateGetBodyRates_f();
+  stabilization_indi_second_order_filter(body_rates, &filtered_rate_2deriv, &filtered_rate_deriv, &filtered_rate, STABILIZATION_INDI_FILT_OMEGA, STABILIZATION_INDI_FILT_ZETA, STABILIZATION_INDI_FILT_OMEGA_R);
+
+  /* attitude error                          */
+  struct Int32Quat att_err;
+  struct Int32Quat* att_quat = stateGetNedToBodyQuat_i();
+  INT32_QUAT_INV_COMP(att_err, *att_quat, stab_att_sp_quat);
+  /* wrap it in the shortest direction       */
+  INT32_QUAT_WRAP_SHORTEST(att_err);
+  INT32_QUAT_NORMALIZE(att_err);
+
+  /* compute the INDI command */
+  attitude_run_indi(stabilization_att_indi_cmd, &att_err);
+
+  stabilization_cmd[COMMAND_ROLL] = stabilization_att_indi_cmd[COMMAND_ROLL];
+  stabilization_cmd[COMMAND_PITCH] = stabilization_att_indi_cmd[COMMAND_PITCH];
+  stabilization_cmd[COMMAND_YAW] = stabilization_att_indi_cmd[COMMAND_YAW];
+
+  /* bound the result */
+  BoundAbs(stabilization_cmd[COMMAND_ROLL], MAX_PPRZ);
+  BoundAbs(stabilization_cmd[COMMAND_PITCH], MAX_PPRZ);
+  BoundAbs(stabilization_cmd[COMMAND_YAW], MAX_PPRZ);
+}
+
+void stabilization_attitude_read_rc(bool_t in_flight, bool_t in_carefree, bool_t coordinated_turn) {
+  struct FloatQuat q_sp;
+#if USE_EARTH_BOUND_RC_SETPOINT
+  stabilization_attitude_read_rc_setpoint_quat_earth_bound_f(&q_sp, in_flight, in_carefree, coordinated_turn);
+#else
+  stabilization_attitude_read_rc_setpoint_quat_f(&q_sp, in_flight, in_carefree, coordinated_turn);
+#endif
+  QUAT_BFP_OF_REAL(stab_att_sp_quat, q_sp);
+}
+
+#define INTEGRATE_RATES(_rates, _derivative, _freq) { \
+_rates->p = _rates->p + _derivative->p/_freq; \
+_rates->q = _rates->q + _derivative->q/_freq; \
+_rates->r = _rates->r + _derivative->r/_freq; \
+}
+
+#define SECOND_ORDER_ACCELERATION(_input, _ddx, _dx, _x, _omega, _omega2, _zeta, _omega_r, _omega2_r) { \
```

Comment:
unless you need this to be a macro, please use a static inline function or directly write it in the second_order_filter as it is only used once anyway...